drivers/media/platform/raspberrypi/rp1_cfe/cfe.c

```diff
@@ -982,6 +983,63 @@ static void cfe_buffer_queue(struct vb2_buffer *vb)
 	spin_unlock_irqrestore(&cfe->state_lock, flags);
 }
 
+static u64 sensor_link_frequency(struct cfe_device *cfe)
+{
+	struct v4l2_mbus_framefmt *source_fmt;
+	struct v4l2_subdev_state *state;
+	struct v4l2_subdev *subdev;
+	const struct cfe_fmt *fmt;
+	struct media_entity *entity;
+	struct media_pad *pad;
+	s64 link_freq;
+
+	state = v4l2_subdev_lock_and_get_active_state(&cfe->csi2.sd);
+	source_fmt = v4l2_subdev_get_pad_format(&cfe->csi2.sd, state,
+						cfe->fe_csi2_channel);
+	fmt = find_format_by_code(source_fmt->code);
```

Comment:
This is probably wrong?  I think I should do this after finding the entity/subdev through the loop below.

Reply:
Maybe. I suppose you want the nearest thing that truthfully declares either a link frequency, or a (pixel format, pixel rate) pair, which does imply putting it inside the loop.

Reply:
Ok, I've done this now, but can revert if we think it's the wrong thing to do in the future.

Reply:
Actually, this is wrong.  We *want* to be looking at the csi entity pad for the format, that's what's connected to the HW after all.  I will revert my update back to the above code.